api/v1alpha1/tls_types.go

```diff
@@ -0,0 +1,59 @@
+// Copyright Envoy Gateway Authors
+// SPDX-License-Identifier: Apache-2.0
+// The full text of the Apache license is available in the LICENSE file at
+// the root of the repo.
+
+package v1alpha1
+
+type TLSSettings struct {
+	// Version details the minimum/maximum TLS protocol version that
+	// should be supported by this listener.
+	// +optional
+	Version *TLSVersions `json:"version,omitempty"`
+
+	// CipherSuites specifies the set of cipher suites supported when
+	// negotiating TLS 1.0 - 1.2. This setting has no effect for TLS 1.3.
+	// +optional
+	CipherSuites []string `json:"ciphers,omitempty"`
+
+	// ECDHCurves specifies the set of supported ECDH curves.
+	// +optional
+	ECDHCurves []string `json:"ecdhCurves,omitempty"`
+
+	// SignatureAlgorithms specifies which signature algorithms the listener should
+	// support.
+	// +optional
+	SignatureAlgorithms []string `json:"signatureAlgorithms,omitempty"`
+
+	// ALPNProtocols supplies the list of ALPN protocols that should be
+	// exposed by the listener. If left empty, ALPN will not be exposed.
+	// +optional
+	ALPNProtocols []string `json:"alpnProtocols,omitempty"`
```

Comment:
I'm not sure that we need to expose `alpn_protocols` to API.  Are there any use cases other than the default `"h2", "http/1.1` ?

Reply:
In some cases we have customers that want to disable ALPN altogether, or disable negotiating HTTP/2 with ALPN.

Generally, there are cases where it makes sense to have a custom implemented protocol. For example, Istio uses a custom ALPN protocol to distinguish between types of traffic (e.g. https://github.com/istio/istio/issues/40680 ).

Reply:
> In some cases we have customers that want to disable ALPN altogether, or disable negotiating HTTP/2 with ALPN.

AFAIK, normally the server side wants to let the user side choose between HTTP 1.1 and HTTP2 for back compatibility. 

> Generally, there are cases where it makes sense to have a custom implemented protocol. For example, Istio uses a custom ALPN protocol to distinguish between types of traffic (e.g. https://github.com/istio/istio/issues/40680 ).

The Istio use case is legitimate, but it's not a Gateway scenario.

I suggest removing this from the API if we don't have a concrete use case right now. We can always add it later, but it will be really hard to remove an existing feature from API.

Reply:
Envoy disables ALPN by default. Envoy Gateway enables ALPN and sets it to `[h2, http/1.1]` by default. It's a reasonable behavior, but perhaps users should have the option to control this parameter. 

In general, other protos are supported in the [IANA registry](https://www.iana.org/assignments/tls-extensiontype-values/tls-extensiontype-values.xhtml#alpn-protocol-ids). In rare cases, it could be necessary to support other protos, e.g. there was a period where `curl` was [spec-compliant](https://github.com/curl/curl/issues/12259) and used `http/1.0` as ALPN for HTTP/1.0. 

Other Gateways like Emissary-Ingress do support configuration of [alpn settings](https://www.getambassador.io/docs/emissary/latest/topics/running/tls#alpn-protocols). The Emissary docs imply that it could be necessary to explicitly set ALPN to `[h2]` for gRPC. I'm not sure if this generally mandatory or required for specific grpc clients. I assume that this was also the considered when ALPN was included in [CTP](https://github.com/envoyproxy/gateway/issues/1846). Maybe @AliceProxy can provide some context here as well.

Reply:
defer to @envoyproxy/gateway-maintainers for suggestions.